File: cse/api-example.js
// EXAMPLE OF GETTING DATA FROM AN API
// const request = new XMLHttpRequest();
// request.open("GET", "http://localhost:5000/api/user");
// request.setRequestHeader("Access-Control-Allow-Credentials", "true");
// request.setRequestHeader("Content-Type", "application/json");
// request.onload = processData;
// request.send();

// function processData() {
//     const response = JSON.parse(request.response);

//     const h1 = document.createElement("h1");
//     // h1.innerText = `Welcome, ${response.data[0].first_name}`;
//     console.log(response);
// }

// EXAMPLE OF SENDING FROM AN API 
function clickToSend() {
    // get data from inputs 
    const inputs = document.getElementsByTagName("input");
    const email = inputs[0].value;
    const password = inputs[1].value; 
    const data = {
        "email": email,
        "password": password
    } 
    const jsonData = JSON.stringify(data);

    // build a request object
    const request = new XMLHttpRequest();

    // open a connection to my api 
    request.open("POST", "http://localhost:5000/api/signin");

    // customise the request 
    request.setRequestHeader("Access-Control-Allow-Credentials", "true");
    request.setRequestHeader("Content-Type", "application/json");
    
    // specify what happens when data arrives 
    request.onload = processRequestToSendDataResponse;
    
    // specify what happens when an error occurs
    request.onerror = processErrorResponse; 

    // send the data 
    request.send(JSON.stringify(data));

    function processRequestToSendDataResponse() {
        const response = JSON.parse(request.response);

        if (request.status == 200) {
            const h1 = document.createElement("h1");
            h1.innerText = `Welcome, ${response.data.user_id}`;
            
            const body = document.getElementsByTagName("body")[0];
            body.appendChild(h1);

            localStorage.setItem("user-id", response.data.user_id);
            sessionStorage.setItem("user-id", response.data.user_id);
        }


    }

    function processErrorResponse() {
        const response = JSON.parse(request.response);
            
        if (request.status == 400) {
            const p = document.createElement("p");
            p.innerText = response.message; 

            const body = document.getElementsByTagName("body")[0];
            body.appendChild(h1);
        }
    }
}

File: cse/api_example.py
import sqlite3
from flask import Flask, request, jsonify
from flask_cors import CORS 

# database interation
def get_user_details_by_email(user_email):
    connection = sqlite3.connect("D:\\CSE-WebApps-2023-2024\\rau-web-apps-programming-2023-2024\\cse\\api_example.db")
    
    query = f"select id, email, password from users where email='{user_email}'"
    
    cursor = connection.cursor()
    user_details = list(cursor.execute(query))
    if len(user_details) != 0:
        user_details = user_details[0]
    connection.close()

    return user_details

def create_user(first_name, last_name, email, password):
    query = f"""INSERT INTO users(first_name, last_name, email, password)
    VALUES('{first_name}', '{last_name}', '{email}', '{password}')"""

    connection = sqlite3.connect("D:\\CSE-WebApps-2023-2024\\rau-web-apps-programming-2023-2024\\cse\\api_example.db")
    cursor = connection.cursor()
    cursor.execute(query)
    connection.commit()
    connection.close()

# api 
app = Flask("example-api")
CORS(app, resources=r'/api/*')

@app.route("/api/users", methods=["POST"])
def users():
    # get the request data
    body = request.json 

    # validation 
    # check that password and retyped password match
    if body["password"] != body["retyped_password"]:
        response = {
            "message": "Failed to sign up. Password mismatch."
        } 
        response = jsonify(response)
        response.headers.add("Access-Control-Allow-Origin", "*")
        return response, 400

    # create a new user in our db 
    create_user(body["first_name"], body["last_name"], body["email"], body["password"])
    
    # get user id and return to client
    user_details = get_user_details_by_email(body["email"])

    response = {
        "data": {
            "user_id": user_details[0]
        }
    }
    response = jsonify(response)
    response.headers.add("Access-Control-Allow-Origin", "*")
    return response, 200    

@app.route("/api/signin", methods=["POST"])
def signin():
    # get request body 
    body = request.json 

    # extract current user details (check if user exists)
    user_details = get_user_details_by_email(body["email"])

    if len(user_details) == 0:
        response = {
            "message": "Invalid user. User missing."
        }
        response = jsonify(response)
        response.headers.add("Access-Control-Allow-Origin", "*")
        return response, 400
    
    if body["password"] != user_details[2]:
        response = {
            "message": "Invalid data provided. Please try again..."
        }
        response = jsonify(response)
        response.headers.add("Access-Control-Allow-Origin", "*")
        return response, 400
    
    response = {
        "data": {
            "user_id": user_details[0]
        }
    }
    response = jsonify(response)
    response.headers.add("Access-Control-Allow-Origin", "*")
    return response, 200

if __name__ == "__main__":
    app.run(debug=True)